k8s/upgrade/src/bin/upgrade-job/helm/upgrade.rs

```diff
@@ -238,109 +237,78 @@ impl HelmUpgrade {
+                }))
             }
             HelmChart::Umbrella if !self.already_upgraded => {
                 // It should be impossible to hit this error.
-                return UmbrellaChartNotUpgraded.fail();
+                UmbrellaChartNotUpgraded.fail()
             }
             HelmChart::Core if self.already_upgraded => {
-                info!(
-                    "Skipping helm upgrade, as the version of the installed helm chart is the same \
-                    as that of this upgrade-job's helm chart"
-                );
+                // Returned HelmUpgradeRunner logs and exits.
+                Ok(Box::pin(async move {
+                    info!(
+                        "Skipping helm upgrade, as the version of the installed helm chart \
+                        is the same as that of this upgrade-job's helm chart"
+                    );
+
+                    Ok(())
+                }))
             }
             HelmChart::Core if !self.already_upgraded => {
                 // It should be impossible to hit this error.
                 let chart_dir = self
                     .core_chart_dir
                     .ok_or(CoreChartUpgradeNoneChartDir.build())?;
 
-                info!("Starting helm upgrade...");
+                // Running 'helm upgrade --dry-run'.
+                let mut dry_run_extra_args: Vec<String> =
+                    self.core_chart_extra_args.clone().unwrap_or_default();
+                dry_run_extra_args.push("--dry-run".to_string());
+                info!("Running helm upgrade dry-run...");
                 self.client
-                    .upgrade(self.release_name, chart_dir, self.core_chart_extra_args)
+                    .upgrade(
+                        self.release_name.as_str(),
+                        chart_dir.as_path(),
+                        Some(dry_run_extra_args),
+                        false,
+                    )
                     .await?;
-                // This file is no longer required after the upgrade command has been executed.
-                self.upgrade_values_file = None;
-
-                info!("Helm upgrade successful!");
-            }
-            _ => {
-                return InvalidHelmUpgrade.fail();
+                info!("Helm upgrade dry-run succeeded!");
+
+                // Returning HelmUpgradeRunner.
+                Ok(Box::pin(async move {
+                    info!("Starting helm upgrade...");
+                    self.client
+                        .upgrade(
+                            self.release_name,
+                            chart_dir.as_path(),
+                            self.core_chart_extra_args,
+                            true,
+                        )
+                        .await?;
+                    info!("Helm upgrade successful!");
+
+                    // This file is no longer required after the upgrade command has been executed.
+                    self.upgrade_values_file = None;
```

Comment:
self will be dropped, so we don't need this

Reply:
Added a `#[allow(dead_code)]`.

Reply:
Why is that required?

Reply:
`self.upgrade_values_file` is a tempfile handle. it is used by the helm binary, but it is not used in the rust code. After upgrade method returns, we None the container, and so the tempfile is garbage-collected. Without this interaction, the file handle is never used. So it causes problems with the linter.

Reply:
Ah I see.. then simply add `#[allow(unused)]` on top of its declaration in the struct, rater than allowing deadcode for the entire struct:
struct {
a: bool,
#[allow(unused)]
b: bool,
}

Reply:
Right, done.